Add std.typecons.Tuple.rename

John-Colvin opened the pull request:
Particularly useful to `map` on to a range of tuples generated from e.g. `zip` or `cartesianProduct`.

JackStouffer commented:
Would this be considered an equivalent to Python's `namedtuple`?

John-Colvin commented:
@JackStouffer Not quite. This is just a convenience function for renaming tuples. `std.typecons.Tuple` already contains the functionality of `namedTuple`

ntrel commented on std/typecons.d:
```diff
@@ -1548,6 +1548,174 @@ unittest
+ * there are members of `t`.
+ */
+template named(names ...)
+if (allSatisfy!(isSomeString, staticMap!(TypeOf, names)))
```
Does `typeof(names)` work instead of staticMap?

John-Colvin replied:
Yes, thanks for the hint

greenify commented on std/typecons.d:
```diff
@@ -1548,6 +1548,167 @@ unittest
+    auto t2Named = t2.named!(["a": "b", "b": "c"]);
+    assert(t2Named.b == 3);
+    assert(t2Named.c == 4);
+
```
Nitpick: afaik this empty line should be removed?

John-Colvin replied:
done (and another one in the test below)

greenify commented on std/typecons.d:
```diff
@@ -1548,6 +1548,167 @@ unittest
+        import std.range : roundRobin, iota;
+        Tuple!(staticMap!(GetItem, aliasSeqOf!(
+                roundRobin(iota(nT), iota(nT, 2*nT))))) tmp;
+        tmp = t;
```
Yet another nitpick: why don't you do the assignment directly in the line above?

John-Colvin replied:
Fixed, using constructor now instead of assignment.

quickfur commented:
ping @andralex 
New addition, needs approval. Thanks!

JakobOvrum commented:
This doesn't seem that different from `tuple!("foo", "bar")(t.expand)`?

wilzbach commented:
> This doesn't seem that different from tuple!("foo", "bar")(t.expand)?

Ping @John-Colvin. Did you read this comment?

andralex commented:
I approve with two requests:

* Make it a member of `Tuple`, it doesn't apply to anything. It would simplify implementation a tad.
* Have it return a `ref` to the original object, it's more flexible.

John-Colvin commented:
@JakobOvrum in the most basic case, yes. But if you look at the examples you'll see it makes things like
* renaming a single member which leaving the others, or
* removing a name

much easier.

Cauterite commented:
May I ask why the template is called `named` instead of something like `renameFields` ?

John-Colvin commented:
@Cauterite I think I'm going to rename it as I change it to a member of Tuple, as per @andralex's advice. `renameFields` seems like a reasonable choice but I haven't thought about it much.

Cauterite commented:
Oh, no problem then.

John-Colvin commented:
@andralex (or anyone else), I'm struggling to work out how to make this return by `ref`.

andralex commented:
@John-Colvin use a trusted cast to pointer to your target type, and return `*that`. Makes sense?

John-Colvin commented:
@andralex yes, that works thanks.

I just remembered why I originally made it a free function:

```D
zip(iota(10), iota(10, 20)).map!(rename!("a", "b"))
```

instead of now:

```D
zip(iota(10), iota(10, 20)).map!(t => t.rename!("a", "b"))
```

I guess it's not really an important difference.

andralex commented:
Interesting, thanks. Probably making it a member remains sensible. Thx!

andralex commented:
Nice execution/ Thanks!

andralex commented:
Auto-merge toggled on

John-Colvin commented:
Just pushed a new commit with a range pipeline example (just in case you didn't see it before setting auto-merge).

andralex commented:
Even better!

John-Colvin commented:
Well seeing as the auto-tester has choked for (presumably) unrelated reasons...

The doc-gen seems to have made a bit of a mess: see http://dtest.thecybershadow.net/artifact/website-e0b2b4609f1a8e4010ccbe9492ab4af3e3fdd2fb-54955d11d5640083599b20017bf0efaa/web/phobos-prerelease/std_typecons.html#.Tuple.rename.2

Am I incorrectly using `LREF` ?

Also, should I take some more care w.r.t. return type qualifiers? I guess it doesn't make much difference seeing as the `expand` field will have the same type(s), which is where all the data is, but perhaps the dropping of the type qualifiers of the struct itself` will confuse people/code?

aG0aep6G commented on std/typecons.d:
```diff
@@ -745,6 +745,189 @@ template Tuple(Specs...)
+        }
+
+        /**
+         * Overload of $(LREF rename) that takes an associative array
```
Needs to be `$(LREF _rename)`. Else "rename" is highlighted because it's the symbol that's being documented. Same below.

andralex replied:
correctamundo

John-Colvin replied:
thanks, that worked, looking fine now

John-Colvin commented:
So I added some extra tests and found what seems to be a dmd bug: https://issues.dlang.org/show_bug.cgi?id=16418

I have also added a workaround for it here.

It seems that @dlang-bot thinks I'm posting fix because I mentioned it in the commit message :/

lodo1995 commented:
> It seems that @dlang-bot thinks I'm posting fix because I mentioned it in the commit message :/

No, the fix column contains an X, not a V, so the dlang bot is correctly considering this only a mention, not a fix.

John-Colvin commented:
anyone got any idea why the win 32 autotesters are failing?

lodo1995 commented:
> Error: more than 32767 symbols in object file

Seriously this limit?

Without really looking too much into this, it looks this PR is causing a growth in the number of symbols instantiated.

wilzbach commented:
Auto-merge toggled off

wilzbach commented:
> Auto-merge toggled off

It's been on the hot list of the auto tester for a while now and thus blocking other PRs.

> anyone got any idea why the win 32 autotesters are failing?

You get:

```
Fatal Error: Out of memory
```

John-Colvin commented:
I don't have a windows machine to test on so this is a bit difficult for me to debug and fix.

John-Colvin commented:
I've moved the unittests to module scope in order to try and get the tests to pass on windows, but that means lack of documentation. Could do with some advice from compiler devs with windows knowledge here. @yebblies @9rnsr @WalterBright ??

John-Colvin commented:
Ok so it now passes on win32, that's good. But that leaves the documentation very much lacking. I guess I could duplicate the unittest inside the ddoc comments, but that's just.... ewww

schveiguy commented:
Related: http://wiki.dlang.org/DIP82

andralex commented:
@John-Colvin: could you please try the solution discussed in the forum?

John-Colvin commented:
Done

andralex commented:
Auto-merge toggled on

John-Colvin commented:
CircleCI failing because it can't find dub?

John-Colvin commented:
Hooray! I wrote the first draft of this function almost 2 years ago, made the pull request 7 months ago, now we're finally here! I feel a blog post coming on... 

Thanks to @andralex for getting back to this and helping make it phobos worthy, also thanks to everyone else involved for helping and nagging me to get it done.

andralex commented:
Thanks for pushing this through. A blog post sounds like a terrific idea.
